File: mat_mul.py
#!/usr/bin/env python

############################
# imports and arg processing
############################

# grab sys and timing modules
import sys
from time import time

# pull user input
if len(sys.argv) < 2 or len(sys.argv) > 4:
    print("Usage: ./mat_mul.py [-v] N [t]")
    print("")
    print("  N: vector system is size 2^N (accept int,float)")
    print("  t: number of threads (optional, accept int)")
    print(" -v: use vectorized matrix multiply (optional)")
    sys.exit(0)

# check if to use vectorized mat mul
vec = sys.argv[1] == '-v'

# arg number for system size and threads
nArg = 2 if vec else 1
tArg = 3 if vec else 2

# set number of threads if provided
userThreads = len(sys.argv) > tArg
if userThreads:
    nt = int(sys.argv[tArg])
else:
    nt = -1

# if vectorized, pass thread number through
# OMP environment variable
if nt > 0 and vec:
    from os import environ
    environ['OMP_NUM_THREADS']=str(nt)

# ^ that has to happen before imports
import joblib as jl
import numpy as np

###################
# system allocation
###################

# set system size
N = int(2.**float(sys.argv[nArg]))

# create random vector and matrix
v = np.random.randn(N,1)
M = np.random.randn(N,N)

#############################
# helper function definitions
#############################

# function for dotting a matrix row and the in-vector
def dot(r):

    # grab reference to matrix and vector
    global M
    global v

    # loop over matrix columns
    total = 0.
    for i in range(N):
        total = total + M[r,i]*v[i]

    # return dot product
    return total

# function for parallel matrix multiplication
def mat_mul(nt):

    # generator for components of dot product
    #
    # each iterate of comps holds a pointer to the
    # dot function and the argument to pass (row index)
    comps = (jl.delayed(dot)(r) for r in range(N))

    # calculate components by forking rows
    out = jl.Parallel(n_jobs=nt)(comps)

    # wrap in numpy object
    return np.array(out)

###############
# actual timing
###############

# min time for testing
Tmin = 1
T = 0

# do mat muls until minimum time is reached
iters = 0
t0 = time()
while T < Tmin:

    # use vectorized or parallelized
    # function as specified
    if not vec:
        w = mat_mul(nt)
    else:
        w = M.dot(v)

    # record time and count iterate
    T = time()-t0
    iters = iters+1

# print output
print("{:g}".format(T/iters))

File: mat_mul.cc
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <random>
#include <omp.h>

/**
 * Performs a matrix multiplication of a random matrix times
 * a random vector of length 2^N, using t threads
 *
 * @param[in] N log_2(system size)
 * @param[in] t number of OMP threads
 */
int main(int argc,char** argv) {

	// check user input
	if (argc < 2 || argc > 3) {
		puts("Usage: ./dot_product N [t]\n");
		puts("  N: vector system is size 2^N (accept int,float)");
		puts("  t: number of OMP threads (optional, accept int)");
		exit(0);
	}

	// set OMP thread num
	int nt=1;

	// only do this part if we've compiled with OpenMP
#if _OPENMP

	if (argc==3) {

		// if we have three args, user has specified nt
		nt = atoi(argv[2]);
	} else {

		// otherwise, use default from OMP_NUM_THREADS
		nt = omp_get_max_threads();
	}
#endif

	// normal distribution mean and standard dev
	static const double mu=0;
	static const double sigma=1;

	// min T for timing purposes
	static const double Tmin=0.5; // half second

	// set up rng and normal distribution
	static const int seed=0; // for reproducibility
	std::default_random_engine rng(seed);
	std::normal_distribution<double> dist(mu,sigma);

	// pull in vector size, allocate two vectors and a matrix
	int N = pow(2,atof(argv[1]));
	double *v = new double[N];
	double *w = new double[N];
	double *M = new double[N*N];

	// fill vectors/matrix with random values
#pragma omp parallel for num_threads(nt)
	for(int i=0;i<N;i++) {
		w[i] = 0;
		v[i] = dist(rng);
		for(int j=0;j<N;j++) M[i+j*N]=dist(rng);
	}

	// perform matrix multiplication
	double t0 = omp_get_wtime(),T=0;
	int iters=0;

	// do matrix multiplies for a speficied time
	while (T<Tmin) {

		// loop over rows of out vector
#pragma omp parallel for num_threads(nt)
		for(int r=0;r<N;r++) {
			for (int c=0;c<N;c++) {
				w[r] += M[c+r*N] * v[c];
			}
		}

		T = omp_get_wtime()-t0; //update time
		iters++;                // count iterations
	}

	// print time per mat mult
	printf("%g\n",T/iters);

	// cleanup vectors, matrix
	delete[] v;
	delete[] w;
	delete[] M;
}
